File: Data Structures & Algorithms/kruskal/submission-0.cpp
class Solution 
{
public:
    int _find(int node, vector<int>& parents)
    {
        while (node != parents[node])
        {
            parents[node] = parents[parents[node]];
            node = parents[node];
        }

        return node;
    }

    bool _union(int n1, int n2, vector<int>& parents, vector<int>& ranks)
    {
        int p1 = _find(n1, parents);
        int p2 = _find(n2, parents);

        if (p1 == p2) return false;

        if (ranks[p1] > ranks[p2])
        {
            parents[p2] = p1;
            ranks[p1] += ranks[p2];
        }
        else
        {
            parents[p1] = p2;
            ranks[p2] += ranks[p1];
        }

        return true;
    }

    int minimumSpanningTree(vector<vector<int>>& edges, int n) 
    {
        vector<int> ranks(n, 1);
        vector<int> parents(n);
        for (int i = 0; i < n; ++i)
        {
            parents[i] = i;
        }

        auto compare = [](vector<int>& a, vector<int>& b)
        {
            return a[2] < b[2];
        };

        sort(edges.begin(), edges.end(), compare);

        int res = 0;
        int numComponents = n;
        for (auto& e : edges)
        {
            if (_union(e[0], e[1], parents, ranks))
            {
                res += e[2];
                numComponents--;
            }
        }
        

        return numComponents == 1 ? res : -1;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)

        res = []
        for i in range(n):
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            lo = i + 1
            hi = n - 1
            while lo < hi:
                currSum = nums[i] + nums[lo] + nums[hi]
                if currSum == 0:
                    res.append([nums[i], nums[lo], nums[hi]])
                    while lo + 1 < hi and nums[lo] == nums[lo + 1]:
                        lo += 1
                    while lo < hi - 1 and nums[hi] == nums[hi - 1]:
                        hi -= 1
                    lo += 1
                    hi -= 1
                elif currSum > 0:
                    hi -= 1
                else:
                    lo += 1
            
        return res
